engine/storage/image/src/main/java/org/apache/cloudstack/storage/image/TemplateServiceImpl.java

```diff
@@ -489,6 +489,10 @@ public void handleTemplateSync(DataStore store) {
                                     s_logger.info("Removing leftover template " + uniqueName + " entry from template store table");
                                     // remove those leftover entries
                                     _vmTemplateStoreDao.remove(tmpltStore.getId());
+                                    // remove from zones
+                                    _vmTemplateZoneDao.deletePrimaryRecordsForTemplate(tmplt.getId());
+                                    // remove template
+                                    _templateDao.remove(tmplt.getId());
```

Comment:
@Spaceman1984 this probably needs checking, what if the templates are OK for other secondary storages within the same zone; and across zones?

Reply:
I'll test with multiple secondary storages

Reply:
You are correct @rhtyd, this code removes the template if there is an error in any of the storages. I have reverted.